x-pack/test/api_integration/apis/code/feature_controls.ts

```diff
@@ -4,67 +4,411 @@
+      expectResponse: expect200,
+    },
+    {
+      url: `/api/code/search/doc?q=starter`,
+      expectForbidden: expect404,
+      expectResponse: expect200,
+    },
+    {
+      url: `/api/code/suggestions/doc?q=starter`,
+      expectForbidden: expect404,
+      expectResponse: expect200,
+    },
+    {
+      url: `/api/code/search/symbol?q=starter`,
+      expectForbidden: expect404,
+      expectResponse: expect200,
+    },
+    {
+      url: `/api/code/suggestions/symbol?q=starter`,
+      expectForbidden: expect404,
+      expectResponse: expect200,
+    },
+  ];
+
+  async function executeRequest(
+    endpoint: string,
+    username: string,
+    password: string,
+    spaceId?: string
+  ) {
+    const basePath = spaceId ? `/s/${spaceId}` : '';
+
+    return await supertest
+      .get(`${basePath}${endpoint}`)
+      .auth(username, password)
+      .set('kbn-xsrf', 'foo')
+      .then((response: any) => ({ error: undefined, response }))
+      .catch((error: any) => ({ error, response: undefined }));
+  }
+
+  async function executeRequests(
+    username: string,
+    password: string,
+    spaceId: string,
+    expectation: 'forbidden' | 'response'
+  ) {
+    for (const endpoint of endpoints) {
+      log.debug(`hitting ${endpoint}`);
+      const result = await executeRequest(endpoint.url, username, password, spaceId);
+      if (expectation === 'forbidden') {
+        endpoint.expectForbidden(result);
+      } else {
+        endpoint.expectResponse(result);
+      }
+    }
+  }
 
   describe('feature controls', () => {
-    // TODO implement tests similar to APM
+    const kibanaUsername = 'kibana_user';
+    const kibanaUserRoleName = 'kibana_user';
+
+    const kibanaUserPassword = `${kibanaUsername}-password`;
+
+    before(async () => {
+      // Import a repository first
+      await security.user.create(kibanaUsername, {
+        password: kibanaUserPassword,
+        roles: [kibanaUserRoleName],
+        full_name: 'a kibana user',
+      });
+
+      await supertest
+        .post(`/api/code/repo`)
+        .auth(kibanaUsername, kibanaUserPassword)
+        .set('kbn-xsrf', 'foo')
+        .send({ url: 'https://github.com/Microsoft/TypeScript-Node-Starter' })
+        .expect(200);
+    });
+
+    after(async () => {
+      // Delete the repository
+      await supertest
+        .delete(`/api/code/repo/github.com/Microsoft/TypeScript-Node-Starter`)
+        .auth(kibanaUsername, kibanaUserPassword)
+        .set('kbn-xsrf', 'foo')
+        .expect(200);
+
+      await security.user.delete(kibanaUsername);
+    });
+
+    it(`Non admin Code user cannot execute delete without all permission`, async () => {
+      const username = 'logstash_read';
+      const roleName = 'logstash_read';
+      const password = `${username}-password`;
+      try {
+        await security.role.create(roleName, {
+          elasticsearch: {
+            indices: [
+              {
+                names: ['.code-*'],
```

Comment:
You shouldn't need to grant index access to the `.code-*` indices. The feature privilege below is sufficient